File: plugins/agentic-sdlc/src/agentic_sdlc/steps/prompt_step.py
"""Prompt step executor."""

from __future__ import annotations

from typing import TYPE_CHECKING

from agentic_sdlc.console import OutputLevel, extract_summary
from agentic_sdlc.progress import WorkflowStatus, update_step_completed, update_step_failed
from agentic_sdlc.runner import run_claude
from agentic_sdlc.steps.base import StepContext, StepExecutor, StepResult

if TYPE_CHECKING:
    from agentic_sdlc.console import ConsoleOutput
    from agentic_sdlc.logging.logger import WorkflowLogger
    from agentic_sdlc.parser import StepDefinition
    from agentic_sdlc.progress import WorkflowProgress


class PromptStepExecutor(StepExecutor):
    """Executes prompt steps by sending prompts to Claude."""

    def execute(
        self,
        step: StepDefinition,
        progress: WorkflowProgress,
        context: StepContext,
        logger: WorkflowLogger,
        console: ConsoleOutput,
    ) -> StepResult:
        """Execute a prompt step."""
        prompt = step.prompt or ""
        template_context = context.build_template_context()

        if context.renderer.has_variables(prompt):
            prompt = context.renderer.render_string(prompt, template_context)

        cwd = context.cwd_override or context.repo_root

        # Load agent file if specified
        if step.agent:
            agent_path = context.repo_root / step.agent
            if agent_path.exists():
                agent_content = agent_path.read_text(encoding="utf-8")
                prompt = f"{agent_content}\n\n{prompt}"

        timeout = (step.step_timeout_minutes or 60) * 60
        max_retry = step.step_max_retry or context.config["defaults"]["maxRetry"]
        bypass_permissions = context.workflow_settings.bypass_permissions if context.workflow_settings else False
        allowed_tools = context.workflow_settings.required_tools if context.workflow_settings else None

        # Always enable streaming when console is provided (handles both BASE and ALL modes)
        print_output = True

        for attempt in range(max_retry + 1):
            result = run_claude(
                prompt=prompt,
                cwd=cwd,
                model=context.resolve_model(step.model),
                timeout=timeout,
                print_output=print_output,
                skip_permissions=bypass_permissions,
                allowed_tools=allowed_tools,
                console=console,
                workflow_id=context.workflow_id,
            )

            if result.success:
                # Use session output context if available, fall back to extracted summary
                session_out = result.session_output
                if session_out.is_success and session_out.context:
                    output_summary = session_out.context
                    if session_out.session_id:
                        logger.info(step.name, f"Session ID: {session_out.session_id}")
                else:
                    output_summary = extract_summary(result.stdout) if result.stdout else ""
                update_step_completed(progress, step.name, output_summary, result.stdout)
                console.step_complete(step.name, output_summary)
                logger.info(step.name, "Step completed successfully")
                return StepResult(success=True, output_summary=output_summary, full_output=result.stdout)

            if attempt < max_retry:
                error_summary = extract_summary(result.stderr) if result.stderr else "Unknown error"
                console.step_retry(step.name, attempt + 1, max_retry + 1, error_summary)
                logger.warning(
                    step.name,
                    f"Attempt {attempt + 1} failed, retrying...",
                    error=result.stderr,
                )
                if progress.current_step:
                    progress.current_step["retry_count"] = attempt + 1
            else:
                error_msg = result.stderr or "Step failed"
                error_summary = extract_summary(error_msg) if error_msg else "Unknown error"
                console.step_failed(step.name, error_summary)
                update_step_failed(progress, step.name, error_msg)
                progress.status = WorkflowStatus.FAILED.value
                logger.error(step.name, f"Step failed after {max_retry + 1} attempts")
                return StepResult(success=False, error=error_msg)

        return StepResult(success=False, error="Unexpected exit from retry loop")

File: src/steps/prompt-step.ts
/** Prompt step executor. */

import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import type { ConsoleOutput } from "../console.js";
import { extractJson, extractSummary } from "../console.js";
import type { WorkflowLogger } from "../logging/logger.js";
import { WORKFLOW_STATUS, updateStepCompleted, updateStepFailed } from "../progress.js";
import { resolveRuntime } from "../runtimes/index.js";
import { runRuntime } from "../runtimes/process-runner.js";
import type { StepDefinition, WorkflowProgress } from "../types.js";
import {
	type StepContext,
	StepExecutor,
	type StepResult,
	buildTemplateContext,
	resolveModel,
} from "./base.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export class PromptStepExecutor extends StepExecutor {
	async execute(
		step: StepDefinition,
		progress: WorkflowProgress,
		context: StepContext,
		logger: WorkflowLogger,
		console: ConsoleOutput,
	): Promise<StepResult> {
		let prompt = step.prompt ?? "";
		const templateContext = buildTemplateContext(context);

		if (context.renderer.hasVariables(prompt)) {
			prompt = context.renderer.renderString(prompt, templateContext);
		}

		const cwd = context.cwdOverride ?? context.repoRoot;

		// Load agent file if specified (fallback to bundled agents)
		if (step.agent) {
			let agentPath = path.join(context.repoRoot, step.agent);
			if (!existsSync(agentPath)) {
				agentPath = path.join(__dirname, "..", "agents", path.basename(step.agent));
			}
			if (existsSync(agentPath)) {
				const agentContent = readFileSync(agentPath, "utf-8");
				prompt = `${agentContent}\n\n${prompt}`;
			}
		}

		const timeout = (step.stepTimeoutMinutes ?? 60) * 60;
		const defaults = context.config.defaults as Record<string, unknown> | undefined;
		const maxRetry = step.stepMaxRetry ?? (defaults?.maxRetry as number) ?? 3;
		const bypassPermissions = context.workflowSettings?.bypassPermissions ?? false;
		const allowedTools = context.workflowSettings?.requiredTools?.length
			? context.workflowSettings.requiredTools
			: null;

		// Always enable streaming when console is provided
		const printOutput = true;

		for (let attempt = 0; attempt <= maxRetry; attempt++) {
			const result = await runRuntime(context.runtimeAdapter, {
				prompt,
				cwd,
				model: resolveModel(context, step.model),
				timeout,
				printOutput,
				skipPermissions: bypassPermissions,
				allowedTools,
				console,
				workflowId: context.workflowId,
				outputDir: context.outputDir,
			});

			if (result.success) {
				// Use session output context if available, fall back to extracted summary
				const sessionOut = result.sessionOutput;
				let outputSummary: string;
				if (sessionOut.isSuccess && sessionOut.context) {
					outputSummary = sessionOut.context;
					if (sessionOut.sessionId) {
						logger.info(step.name, `Session ID: ${sessionOut.sessionId}`);
					}
				} else {
					outputSummary = result.stdout ? extractSummary(result.stdout) : "";
				}

				// Try to extract structured JSON from skill output
				const stepOutput = result.stdout ? extractJson(result.stdout) : null;
				const outputValue = stepOutput ?? result.stdout;

				updateStepCompleted(progress, step.name, outputSummary, outputValue, sessionOut.sessionId);
				console.stepComplete(step.name, outputSummary);
				logger.info(step.name, "Step completed successfully");
				return { success: true, outputSummary, fullOutput: result.stdout };
			}

			if (attempt < maxRetry) {
				const errorSummary = result.stderr ? extractSummary(result.stderr) : "Unknown error";
				console.stepRetry(step.name, attempt + 1, maxRetry + 1, errorSummary);
				logger.warning(step.name, `Attempt ${attempt + 1} failed, retrying...`, {
					error: result.stderr,
				});
				if (progress.currentStep) {
					progress.currentStep.retryCount = attempt + 1;
				}
			} else {
				const errorMsg = result.stderr || "Step failed";
				const errorSummary = errorMsg ? extractSummary(errorMsg) : "Unknown error";
				console.stepFailed(step.name, errorSummary);
				updateStepFailed(progress, step.name, errorMsg);
				progress.status = WORKFLOW_STATUS.FAILED;
				logger.error(step.name, `Step failed after ${maxRetry + 1} attempts`);
				return { success: false, error: errorMsg };
			}
		}

		return { success: false, error: "Unexpected exit from retry loop" };
	}
}
